src/Streamly/FileSystem/Event.hs

```diff
@@ -0,0 +1,109 @@
+-- |
+-- Module      : Streamly.FileSystem.Event
+-- Copyright   : (c) 2020 Composewell Technologies
+--
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+
+module Streamly.FileSystem.Event
+    ( Config
+    , Event
+    , Toggle
+    , defaultConfig
+    , setAllEvents
+
+    -- ** Events of Interest
+    , setRootMoved
+
+    -- ** Watch APIs
+    , watch
+
+    -- * Handling Events
+    , getAbsolutePath
+
+    -- ** Item CRUD events
+    , isCreated
+    , isDeleted
+    , isMoved
+    , isModified
+
+    -- ** Exception Conditions
+    , isOverflow
+
+    -- * Debugging
+    , showEvent
+    )
+where
+
+import Data.List.NonEmpty (NonEmpty)
+import Data.Word (Word8)
+import Streamly.Prelude (SerialT)
+import Streamly.Internal.Data.Array.Storable.Foreign (Array)
+
+#if defined(CABAL_OS_DARWIN)
+import qualified Streamly.Internal.FileSystem.Event.Darwin as Event
+#elif defined(CABAL_OS_LINUX)
+import qualified Streamly.Internal.FileSystem.Event.Linux as Event
+#elif defined(CABAL_OS_WINDOWS)
+import qualified Streamly.Internal.FileSystem.Event.Windows as Event
+#else
+#error "FS Events not supported on this platform
+#endif
+
+-------------------------------------------------------------------------------
+-- Main
+-------------------------------------------------------------------------------
+type Toggle = Event.Toggle
+type Config = Event.Config
+type Event = Event.Event
+
+setRootMoved :: Toggle -> Config -> Config
+setRootMoved = Event.setRootMoved
+
+setAllEvents :: Toggle -> Config -> Config
+#if defined(CABAL_OS_DARWIN)
+setAllEvents = defaultConfig
```

Comment:
I think there will be a type mismatch here? I wonder why the CI for mac does not fail.
`defaultConfig :: Config`
`setAllEvents :: Toggle -> Config -> Config`